Microphone support

jroweboy opened the pull request:
This adds microphone service implementations using both a real microphone and a static microphone. As i'm not too sure about the design, I pushed it working and would like to ask for any feedback before cleaning up and getting it merged. It should be canary ready from the get go if someone wants to tag it.

Design:
Frontend selects a Mic implementation. Core will "get" the current Mic implementation when loading MIC_U services. When the game will configure the parameters for sampling, core will forward these to the frontend mic impl. Each frontend mic is responsible for collecting whatever data it needs, and managing its threads. The MIC_U service will start a CoreTiming event loop that will call Read on the frontend Mic impl and get the latest data from the frontend. This is sync and blocks the core emulation, so frontends should take care to do as little as possible at this point.

In general the code should be decent quality, but I am already bored of working on it again, so I want to get it out. Requesting for any design decisions before I get the standard c++ code quality treatment, so i don't have to waste much time if the whole thing needs a rewrite anyway.

Real microphone runs on a separate thread using cubeb input support and writes the data into a spsc queue. When read is called, it will pull as much as it can from the queue and write it to shared mem.

Static microphone returns the same N random bytes every time Read is called.

Thanks to tobi for fixing that bug i had and pressuring me to finally finish this.

THIS DOES NOT INCLUDE A KEY OR CONTROLLER MAPPING TO LET YOU "PUSH TO TALK" Someone else can add that later if they want. shouldn't be too hard.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4671)
<!-- Reviewable:end -->

adityaruplaha commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -14,24 +15,106 @@
+    PCM8 = 0,        ///< Unsigned 8-bit PCM.
+    PCM16 = 1,       ///< Unsigned 16-bit PCM.
+    PCM8Signed = 2,  ///< Signed 8-bit PCM.
+    PCM16Signed = 3, ///< Signed 16-bit PCM.
```
Are `= 0` etc. necessary? Ditto below.

lioncash commented on src/audio_core/cubeb_input.cpp:
```diff
@@ -0,0 +1,145 @@
+long CubebInput::Impl::DataCallback(cubeb_stream* stream, void* user_data, const void* input_buffer,
+                                    void* output_buffer, long num_frames) {
+    Impl* impl = static_cast<Impl*>(user_data);
+    u8 const* data = reinterpret_cast<u8 const*>(input_buffer);
```
```suggestion
    const u8* data = reinterpret_cast<const u8*>(input_buffer);
```

This can also be moved after the conditional check below.

lioncash commented on src/core/frontend/mic.cpp:
```diff
@@ -0,0 +1,86 @@
+    return (sample_size == 8) ? CACHE_8_BIT : CACHE_16_BIT;
+}
+
+static std::shared_ptr<Mic::Interface> current_mic;
```
This should probably go into the system class.

lioncash commented on src/core/frontend/mic.h:
```diff
@@ -0,0 +1,126 @@
+        return is_sampling;
+    }
+
+    Parameters GetParameters() const {
```
It may be worthwhile to return by const reference here, which avoids always copying the data except when explicitly desired.

lioncash commented on src/core/hle/service/mic_u.h:
```diff
@@ -192,4 +192,6 @@ class MIC_U final : public ServiceFramework<MIC_U> {
 
 void InstallInterfaces(Core::System& system);
 
+void ChangeMicImpl();
```
As far as I can tell, this doesn't seem to be implemented anywhere.

lioncash commented on src/core/settings.h:
```diff
@@ -167,6 +167,8 @@ struct Values {
     bool enable_audio_stretching;
     std::string audio_device_id;
     float volume;
+    u8 mic_input_type;
```
It may be preferable here to use an enum (which is coerced to/from a u8 when writing/reading the value.

legend800 commented:
Just tested static and real mic and everything functioned correctly with SM3DL and WarioWare Gold. Amazing work!

zhaowenlan1779 commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -14,24 +15,106 @@
             system.Kernel().CreateEvent(Kernel::ResetType::OneShot, "MIC_U::buffer_full_event");
+        buffer_write_event = timing.RegisterEvent(
+            "MIC_U::UpdateBuffer", std::bind(&Impl::UpdateSharedMemBuffer, this,
+                                             std::placeholders::_1, std::placeholders::_2));
```
Might as well consider using lambdas to make this easier to read

zhaowenlan1779 commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -14,24 +15,106 @@
+    u8 sample_size = 0;
+    SampleRate sample_rate = SampleRate::Rate16360;
+
+    void UpdateOffset() {
```
Can't this just be moved into `WriteSamples`? (or at least call this there) As far as I can tell this is not called anywhere else and moving it can help prevent people forgetting to call this.

wwylele commented on src/core/frontend/mic.cpp:
```diff
@@ -0,0 +1,86 @@
+void StaticMic::AdjustSampleRate(u32 sample_rate) {}
+
+Samples StaticMic::Read() {
+    return (sample_size == 8) ? CACHE_8_BIT : CACHE_16_BIT;
```
Isn't this too large, as MIC service calls this function every 15 (16?) samples?

wwylele commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
         rb.Push(RESULT_SUCCESS);
-        is_sampling = false;
-        LOG_WARNING(Service_MIC, "(STUBBED) called");
+        mic->StopSampling();
```
As `timing.ScheduleEvent` is called in `StartSampling`, I'd expect `CancelEvent` is called here symmetrically.

wwylele commented:
Please fix compiler error before we can merge this into canary.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
         rb.Push(RESULT_SUCCESS);
 
-        LOG_WARNING(Service_MIC, "called, size=0x{:X}", size);
+        LOG_TRACE(Service_MIC, "MIC:U MapSharedMem called, size=0x{:X}", size);
```
no need to include function name here.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
         shared_memory = nullptr;
         rb.Push(RESULT_SUCCESS);
-        LOG_WARNING(Service_MIC, "called");
+        LOG_TRACE(Service_MIC, "MIC:U UnmapSharedMem called");
```
dito.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
-                    static_cast<u32>(encoding), static_cast<u32>(sample_rate), audio_buffer_offset,
-                    audio_buffer_size, audio_buffer_loop);
+        LOG_TRACE(Service_MIC,
+                  "MIC:U StartSampling called, encoding={}, sample_rate={}, "
```
dito

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
         IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
         rb.Push(RESULT_SUCCESS);
-        LOG_WARNING(Service_MIC, "(STUBBED) called, sample_rate={}", static_cast<u32>(sample_rate));
+        LOG_TRACE(Service_MIC, "MIC:U AdjustSampling sample_rate={}",
```
dito

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
+        bool is_sampling = mic->IsSampling();
         rb.Push<bool>(is_sampling);
-        LOG_WARNING(Service_MIC, "(STUBBED) called");
+        LOG_TRACE(Service_MIC, "MIC:U IsSampling: {}", is_sampling);
```
dito

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -111,37 +241,42 @@ struct MIC_U::Impl {
         IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
         rb.Push(RESULT_SUCCESS);
-        LOG_WARNING(Service_MIC, "(STUBBED) called, mic_gain={}", mic_gain);
+        LOG_TRACE(Service_MIC, "MIC:U SetGain gain={}", gain);
```
dito

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +124,115 @@ struct MIC_U::Impl {
-        is_sampling = false;
-        LOG_WARNING(Service_MIC, "(STUBBED) called");
+        mic->StopSampling();
+        LOG_TRACE(Service_MIC, "MIC:U StopSampling called");
```
dito.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -111,37 +241,42 @@ struct MIC_U::Impl {
-        LOG_WARNING(Service_MIC, "(STUBBED) called");
+        u8 gain = mic->GetGain();
+        rb.Push<u8>(gain);
+        LOG_TRACE(Service_MIC, "MIC:U GetGain gain={}", gain);
```
dito.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -111,37 +241,42 @@ struct MIC_U::Impl {
         IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
         rb.Push(RESULT_SUCCESS);
-        LOG_WARNING(Service_MIC, "(STUBBED) called, mic_power={}", mic_power);
+        LOG_TRACE(Service_MIC, "MIC:U SetPower mic_power={}", power);
```
dito.

FearlessTobi commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -111,37 +241,42 @@ struct MIC_U::Impl {
+        bool mic_power = mic->GetPower();
         rb.Push<u8>(mic_power);
-        LOG_WARNING(Service_MIC, "(STUBBED) called");
+        LOG_TRACE(Service_MIC, "MIC:U GetPower called");
```
dito.

jroweboy commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -14,24 +15,106 @@
+        size_t bytes_to_write = std::min(samples.size(), remaining_space);
+
+        // Write as many samples as we can to the buffer.
+        // TODO if the sample size is 16bit, this could theoretically cut a sample
```
its not going to sharedmem size, but to the "size" sampling parameter that the application configured. meaning the game *could* set size to an odd value, but there isn't a good reason for them to do that.

jroweboy commented:
I think i've addressed all the feedback. Let me know if i missed your suggestion, there was a lot.

wwylele commented on src/core/frontend/mic.cpp:
```diff
@@ -0,0 +1,62 @@
+
+namespace Frontend::Mic {
+
+constexpr std::array<u8, 32> NOISE_SAMPLE_8_BIT = {
```
I'd expect this buffer only has 16 bytes (half of the 16 bit one)

wwylele commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -14,24 +15,106 @@
+constexpr u64 BufferUpdateRate8180 = BASE_CLOCK_RATE_ARM11 / 511;
+constexpr u64 BufferUpdateRate10910 = BASE_CLOCK_RATE_ARM11 / 681;
+constexpr u64 BufferUpdateRate16360 = BASE_CLOCK_RATE_ARM11 / 1022;
+constexpr u64 BufferUpdateRate32730 = BASE_CLOCK_RATE_ARM11 / 2045;
```
Leftover code you forgot to remove?

wwylele commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -41,68 +115,113 @@ struct MIC_U::Impl {
-                    static_cast<u32>(encoding), static_cast<u32>(sample_rate), audio_buffer_offset,
-                    audio_buffer_size, audio_buffer_loop);
+        LOG_TRACE(Service_MIC,
+                  "StartSampling called, encoding={}, sample_rate={}, "
```
fearlessTobi meant that you don't need `StartSampling`. Just write `"called, ..."` because the function name is automatically logged by the logging system.

wwylele commented on src/audio_core/cubeb_input.cpp:
```diff
@@ -0,0 +1,151 @@
+        return 0;
+    }
+
+    u8 const* data = reinterpret_cast<u8 const*>(input_buffer);
```
Still, `const u8*` to be consistent with our code style.

jroweboy replied:
that changes the semantics. it would be a `const u8` pointer instead of `const pointer` to a `u8`

wwylele commented on src/core/frontend/mic.h:
```diff
@@ -0,0 +1,118 @@
+    virtual Samples Read() = 0;
+
+    /// Adjusts the Parameters. Implementations should update the parameters field in addition to
+    /// changing the mic to sample according to the new parameters. Called by Core
```
Another multiline doc here

wwylele commented on src/core/hle/service/mic_u.cpp:
```diff
@@ -292,14 +412,33 @@ MIC_U::MIC_U(Core::System& system)
 } // namespace Service::MIC
+
+namespace Frontend::Mic {
+static std::shared_ptr<Mic::Interface> current_mic;
```
This is still here as a global state. You can partially resolve this by:
 - only keeps a unique_ptr in the MIC service class
 - construct MIC::Interface in MIC service constructor according to settings value
 - Settings Apply notifies MIC service to reconstruct MIC::Interface if you want hot select

jroweboy commented:
Changes since last update:
Removed global state/ addressed all review feedback
Add mic hot swapping (so you can change mic settings while the game is running seamlessly)
Should've fixed android compiling support (no real mic support for android atm)

zhaowenlan1779 commented on src/core/settings.cpp:
```diff
@@ -5,10 +5,12 @@
 #include <utility>
 #include "audio_core/dsp_interface.h"
 #include "core/core.h"
+#include "core/frontend/emu_window.h"
```
where is this used?

zhaowenlan1779 commented on src/core/frontend/mic.cpp:
```diff
@@ -0,0 +1,61 @@
+
+#include <array>
+#include "core/frontend/mic.h"
+#include "core/hle/service/mic_u.h"
```
where is this used?

slashiee commented:
Hmm... my microphone seems to not be cooperating with this feature?
https://youtu.be/H_iqTS2Q8gY

jroweboy commented:
@slashiee please retest with canary 1260 or greater. if its still not working, please paste a log and a save file for it (if needed) so we can replicate the issue.

Changes since last time
* Addressed review comments
* Added support for 8bit pcm samples in cubeb (real device) input.

I haven't tested 8bit yet, but its pretty straight forward. If its broken, i'll go the extra mile to get that tested as well.

slashiee commented:
It's working now, thank you

jroweboy commented:
CI is unrelated.

The macOS crash is verified fixed https://community.citra-emu.org/t/real-mic-input-causes-canary-to-crash-on-osx-10-14/93053/4

Ready for mergeeeeeeeeeeee?

NeilZander commented:
I'm sorry for acting like a noob. Not entirely sure where to go with finding things on github... I'm having trouble finding where the download is so I can finally use microphone support? or am I sadly reading this wrong and it's not available yet?

slashiee commented:
It's already a part of nightly and Canary.

NeilZander commented:
ah okay thanks for replying one last question then does it activate automatically? When I downloaded the update I could not find anything to show it was actually there that's why I could not tell if I already had it or not.

jroweboy commented:
@NeilZander its in configure system audio settings. But also this is an issue tracker for developers and not a support forum. Please use the forums for support requests. https://community.citra-emu.org/ Go there if you need more information, thanks.
